File: ios/AssetManager/ios_Latest_Apr_24/ios/AssetManager/AssetManager/PurchaseOrder.m

//////////////////////////////////////////////////////////////////////////////
//This file is part of Cireson Barcode Scanner. 
//
//Cireson Barcode Scanner is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//(at your option) any later version.
//
//Cireson Barcode Scanner is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with Cireson Barcode Scanner.  If not, see<https://www.gnu.org/licenses/>.
/////////////////////////////////////////////////////////////////////////////
//
//  PurchaseOrder.m
//  AssetManager
//

#import "PurchaseOrder.h"

@implementation PurchaseOrder

@synthesize orderNumber, totalCost, purchaseOrderDate, purchaseOrderStatus, associatedAssets;

-(BOOL)load:(NSDictionary*)obj {
    BOOL result = [super load:obj];
    if (result) {
        self.associatedAssets = Nil; // reset the current asset list.
        NSArray* assets = [obj objectForKey:@"Source_HardwareAssetHasPurchaseOrder"];
        if (Nil != assets && [assets count] > 0) {
            NSMutableArray* tempAssets = [[NSMutableArray alloc] init];
            for(NSDictionary* assetObj in assets) {
                Asset* asset = [[Asset alloc] init];
                if ([asset load:assetObj]) {
                    [tempAssets addObject:asset];
                }
            }
            self.associatedAssets = tempAssets;
        }
        result = YES;
    }
    return result;
}

+(PurchaseOrder*)loadFromDict:(NSDictionary*)dict {
    PurchaseOrder* po = [[PurchaseOrder alloc] init];
    if ([po load:dict])
        return po;
    return Nil;
}

+(NSArray*)loadFromArray:(NSArray*)arr {
    if (arr == Nil) return Nil;
    NSMutableArray* tmpArr = [[NSMutableArray alloc] init];
    for(NSDictionary* poObj in arr) {
        PurchaseOrder* po = [PurchaseOrder loadFromDict:poObj];
        if (po != Nil) {
            [tmpArr addObject:po];
        }
    }
    return tmpArr;
}

-(NSString*)orderNumber {
    return [self readStringFromObject:self.jsonObject key:@"PurchaseOrderNumber"];
}

-(NSNumber*)totalCost {
    return [self readNumberFromObject:self.jsonObject key:@"Amount"];
}

-(NSString*)displayName {
    return [self readStringFromObject:self.jsonObject key:@"DisplayName"];
}

-(NSString*)purchaseOrderDate {
    return [self readStringFromObject:self.jsonObject key:@"PurchaseOrderDate"];
}

-(NSString*)purchaseOrderStatus {
    NSDictionary* statusObj = [self.jsonObject objectForKey:@"PurchaseOrderStatus"];
    return [self readStringFromObject:statusObj key:@"Name"];
}

-(void)updateAssets {
    
    NSMutableArray* jsonArray = [[NSMutableArray alloc] init];
    
    // create json array
    for (Asset* a in self.associatedAssets) {
        [jsonArray addObject:a.jsonObject];
    }
    
    // because this is a contained object we also need to udpate the current json
    [self.jsonObject setValue:jsonArray forKey:@"Source_HardwareAssetHasPurchaseOrder"];
}

@end

File: android/AssetManager/src/app/src/main/java/com/cireson/assetmanager/model/PurchaseOrder.java

//////////////////////////////////////////////////////////////////////////////
//This file is part of Cireson Barcode Scanner. 
//
//Cireson Barcode Scanner is free software: you can redistribute it and/or modify
//it under the terms of the GNU General Public License as published by
//the Free Software Foundation, either version 3 of the License, or
//(at your option) any later version.
//
//Cireson Barcode Scanner is distributed in the hope that it will be useful,
//but WITHOUT ANY WARRANTY; without even the implied warranty of
//MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
//GNU General Public License for more details.
//
//You should have received a copy of the GNU General Public License
//along with Cireson Barcode Scanner.  If not, see<https://www.gnu.org/licenses/>.
/////////////////////////////////////////////////////////////////////////////


package com.cireson.assetmanager.model;

import com.cireson.assetmanager.service.CiresonCriteria;
import com.cireson.assetmanager.service.BaseJsonObject;
import com.cireson.assetmanager.util.BooleanSerializer;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;
import com.google.gson.annotations.Expose;
import com.google.gson.annotations.SerializedName;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;

/**
 * Created by welcome on 4/9/14.
 */
public class PurchaseOrder extends BaseJsonObject {

    /*Serializable Instances..*/
    @SerializedName("Amount")
    public float amount = 0;
    @SerializedName("DisplayName")
    public String displayName = null;
    @SerializedName("PurchaseOrderDate")
    public String purchaseOrderDate = null;
    @SerializedName("ClassTypeId")
    public String classTypeId = null;
    @SerializedName("ObjectStatus")
    public CiresonObjectStatus objectStatus = null;
    @SerializedName("Source_HardwareAssetHasPurchaseOrder")
    public ArrayList<Assets> associatedAssets = null;
    @SerializedName("Target_PurchaseOrderHasCostCenter")
    public Object targetPurchaseOrderHasCostCenter = null;
    @SerializedName("PurchaseOrderStatus")
    public PurchaseOrderStatus purchaseOrderStatus = null;

    public static class PurchaseOrderStatus{
        @Expose
        @SerializedName("Id")
        public String id = null;
        @Expose
        @SerializedName("Name")
        public String name = null;
    }

    public PurchaseOrder(){/*Default..*/}

    /*Required Inner Classes..*/
    /*This gson helps in serializing and deserializing PurchaseOrder model..*/
    private static transient Gson gson;
    private JsonElement originalJson;
    /*Custom serializer and deserializer for PurchaseOrder model.*/
     /*Custom JSON Serializer and deserializer of Assets model..*/
    static{
        gson = new GsonBuilder()
                .setExclusionStrategies(new MyExclusionStrategy())
                .registerTypeAdapter(Boolean.class, new BooleanSerializer())
                .registerTypeAdapter(CiresonCriteria.SimpleExpression.class, new CiresonCriteria.SimpleExpressionSerializer())
                .create();
    }

    public static class OriginalJsonSerializer implements JsonSerializer<PurchaseOrder> {
        @Override
        public JsonElement serialize(PurchaseOrder o, Type type, JsonSerializationContext jsonSerializationContext) {
            if(o.originalJson!=null){
                JsonObject object = o.originalJson.getAsJsonObject();
                if(object!=null){
                    if(object.has("Source_HardwareAssetHasPurchaseOrder")){
                        JsonArray newJsonArray = new JsonArray();
                        for(Assets a: GlobalData.getInstance().getSavedReceiveAssets().projectionObject.associatedAssets){
                            JsonObject newAssetObject = a.originalJson.getAsJsonObject();
                            newJsonArray.add(newAssetObject);
                            object.add("Source_HardwareAssetHasPurchaseOrder",newJsonArray);
                        }
                    }
                }
            }
            return o.originalJson;
        }
    }

    public static class OriginalJsonDeserializer implements JsonDeserializer<PurchaseOrder> {
        @Override
        public PurchaseOrder deserialize(JsonElement jsonElement, Type type, JsonDeserializationContext jsonDeserializationContext) throws JsonParseException {
            PurchaseOrder purchaseOrder = gson.fromJson(jsonElement,type);
            purchaseOrder.originalJson = jsonElement;
            if(purchaseOrder.originalJson!=null){
                JsonObject jsonObject = purchaseOrder.originalJson.getAsJsonObject();
                final Type classType = new TypeToken<ArrayList<Assets>>(){}.getType();
                ArrayList<Assets> associatedAssets = jsonDeserializationContext.deserialize(jsonObject.get("Source_HardwareAssetHasPurchaseOrder"), classType);
                purchaseOrder.associatedAssets = associatedAssets;
            }
            return purchaseOrder;
        }
    }

    /*Getters and setters..*/

}
